pkg/offers/offers_test.go

```diff
@@ -47,6 +73,212 @@ func TestTimedOffer(t *testing.T) {
+	o = &mesos.Offer{Id: id, Hostname:&incompatibleHostname}
+	declinedNumBefore = declinedNum
+	storage.Add([]*mesos.Offer{o})
+	if obj, ok := storage.Get(id.GetValue()); obj != nil || ok {
+		t.Error("incompatible offer not rejected")
+	}
+	if declinedNum <= declinedNumBefore {
+		t.Error("incompatible offer is not declined")
+	}
+
+	// Invalidated offer are not declined, but expired
+	id = util.NewOfferID("foo5")
+	o = &mesos.Offer{Id: id}
+	storage.Add([]*mesos.Offer{o})
+	offer, _ = storage.Get(id.GetValue())
+	declinedNumBefore = declinedNum
+	storage.Invalidate(id.GetValue())
+	if obj, _ := storage.Get(id.GetValue()); !obj.HasExpired() {
+		t.Error("invalidated offer is not expired")
+	}
+	if declinedNum > declinedNumBefore {
+		t.Error("invalidated offer is declined")
+	}
+	if ok := offer.Acquire(); ok {
+		t.Error("invalidated offer can be acquired")
+	}
+
+	// Invalidate "" will invalidate all offers
+	id = util.NewOfferID("foo6")
+	o = &mesos.Offer{Id: id}
+	storage.Add([]*mesos.Offer{o})
+	id2 := util.NewOfferID("foo7")
+	o2 := &mesos.Offer{Id: id2}
+	storage.Add([]*mesos.Offer{o2})
+	storage.Invalidate("")
+	if obj, _ := storage.Get(id.GetValue()); !obj.HasExpired() {
+		t.Error("invalidated offer is not expired")
+	}
+	if obj2, _ := storage.Get(id2.GetValue()); !obj2.HasExpired() {
+		t.Error("invalidated offer is not expired")
+	}
+
+	// InvalidateForSlave invalides all offers for that slave, but only those
+	id = util.NewOfferID("foo8")
+	slaveId := util.NewSlaveID("test-slave")
+	o = &mesos.Offer{Id: id, SlaveId:slaveId}
+	storage.Add([]*mesos.Offer{o})
+	id2 = util.NewOfferID("foo9")
+	o2 = &mesos.Offer{Id: id2}
+	storage.Add([]*mesos.Offer{o2})
+	storage.InvalidateForSlave(slaveId.GetValue())
+	if obj, _ := storage.Get(id.GetValue()); !obj.HasExpired() {
+		t.Error("invalidated offer for test-slave is not expired")
+	}
+	if obj2, _ := storage.Get(id2.GetValue()); obj2.HasExpired() {
+		t.Error("invalidated offer another slave is expired")
+	}
+
+	close(done)
+} // TestOfferStorage
+
+func TestListen(t *testing.T) {
+	ttl := time.Second / 4
+	config := RegistryConfig{
+		DeclineOffer: func(offerId string) <-chan error {
+			return proc.ErrorChan(nil)
+		},
+		Compat: func(o *mesos.Offer) bool {
+			return true
+		},
+		TTL:           ttl,
+		ListenerDelay: ttl / 2,
+	}
+	storage := CreateRegistry(config)
+
+	done := make(chan struct {})
+	storage.Init(done)
+
+	// Create two listeners with a hostname filter
+	hostname1 := "hostname1"
+	hostname2 := "hostname2"
+	listener1 := storage.Listen("listener1", func(offer *mesos.Offer) bool {
+		return offer.GetHostname() == hostname1
+	})
+	listener2 := storage.Listen("listener2", func(offer *mesos.Offer) bool {
+		return offer.GetHostname() == hostname2
+	})
+
+	// Add hostname1 offer
+	id := util.NewOfferID("foo")
+	o := &mesos.Offer{Id: id, Hostname:&hostname1}
+	storage.Add([]*mesos.Offer{o})
+
+	// listener1 is notified by closing channel
+	select {
+	case _, more := <-listener1:
```

Comment:
(very) small nit: "more" is an interesting choice. i think that if <-someChan returns `_, true` then it doesn't necessarily mean that there's more data coming down the chan, but that it just hasn't closed yet. people tend to use `ok` or `found` for instead.